warehouse/tuf/services.py

```diff
@@ -0,0 +1,378 @@
+        # 3. Bump Snapshot role (and write to the Storage).
+        # 4. Bump Timestamp role using the new Snapshot Metadata.
+
+        # 1. Metadata Repository.
+        metadata_repository = MetadataRepository(
+            self._storage_backend, self._key_storage_backend
+        )
+
+        # 2. Snapshot role metadata.
+        snapshot_metadata = metadata_repository.load_role(Role.SNAPSHOT.value)
+
+        # 3. Bump Snapshot role metadata.
+        snapshot_metadata = metadata_repository.snapshot_bump_version(
+            snapshot_expires=self._set_expiration_for_role(
+                self._request, Role.SNAPSHOT.value
+            ),
+            store=True,
+        )
+
+        # 4. Bump Snapshot role metadata using Timestamp metadata.
+        metadata_repository.timestamp_bump_version(
+            snapshot_version=snapshot_metadata.signed.version,
+            timestamp_expires=self._set_expiration_for_role(
+                self._request, Role.TIMESTAMP.value
+            ),
+            store=True,
+        )
+
+    def bump_bin_n_roles(self):
+        # Bumping all of the delegated bin roles in the TUF repository involves
+        # the following steps:
+        # 1. Metadata Repository.
+        # 2. Load Snapshot role.
+        # 3. Load BINS role.
+        # 4. For each BIN-N role (hash bin) in BINS, fetch the
+        #    role, bump, write back to the repo and update Snapshot role MetaFile.
+        # 5. Bump Snapshot and write back to repository.
+        # 6. Bump Timestamp role (using updated Snapshot) and write back to
+        #    repository.
+
+        # 1. Metadata Repository
+        metadata_repository = MetadataRepository(
+            self._storage_backend, self._key_storage_backend
+        )
+
+        # 2. Load Snapshot role.
+        snapshot_metadata = metadata_repository.load_role(Role.SNAPSHOT.value)
+
+        # 3 Load BINS role.
+        bins_metadata = metadata_repository.load_role(Role.BINS.value)
+
+        # 4. Fore each delegated hashed bin target role, bump and update Snapshot
+        for role in bins_metadata.signed.delegations.roles.keys():
+            role_metadata = metadata_repository.load_role(role)
+            metadata_repository.bump_role_version(
+                rolename=role,
+                role_metadata=role_metadata,
+                role_expires=self._set_expiration_for_role(Role.BINS.value),
+                key_rolename=Role.BIN_N.value,
+                store=True,
+            )
+
+            snapshot_metadata = metadata_repository.snapshot_update_meta(
+                role, role_metadata.signed.version, snapshot_metadata
+            )
+
+        # 5. Bump Snapshot with updated metadata
+        snapshot_metadata = metadata_repository.snapshot_bump_version(
+            snapshot_expires=self._set_expiration_for_role(
+                self._request, Role.SNAPSHOT.value
+            ),
+            snapshot_metadata=snapshot_metadata,
```

Comment:
https://github.com/lukpueh/warehouse/pull/6#discussion_r817528819 applies here too.

Reply:
As @kairoaraujo explained to me, it does not apply here, as we need to pass the `snapshot_metadata` that we incrementally updated in the loop  (in memory only) to `snapshot_bump_version` where we eventually persist all changes at once.